Improve page load performance in rustdoc

jsha opened the pull request:
Add an explicit height to icons (which already had an explicit width) to allow browsers to lay out the page more accurately before the icons have been loaded. https://web.dev/optimize-cls/.

Add min-width: 115px to the crate search dropdown. When the HTML first loads, this dropdown includes only the text "All crates." Later, JS loads the items underneath it, some of which are wider. That causes the dropdown to get wider, causing a distracting reflow. This sets a min-width based on the size that the dropdown eventually becomes based on the crates on doc.rust-lang.org, reducing page movement during load.

Add font-display: swap. Per https://web.dev/font-display/, this prevents "flash of invisible text" during load by using a system font until the custom font is available. I've noticed this flash of invisible text occasionally when reading Rust docs. Note that users without cached fonts will see text, and then see it reflow. For `docs.rust-lang.org`, [setting caching headers will help a lot](https://github.com/rust-lang/simpleinfra/issues/62).

Generated output at https://jacob.hoffman-andrews.com/rust/flow-improvements/std/string/struct.String.html.

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @ollie27 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

the8472 commented:
> Inline the styles for custom fonts so the browser can start loading them earlier.

Wouldn't `<link rel=preload ...>` do the job too?

GuillaumeGomez commented:
Overall, it's all very nice changes. My only concern here is about inling the font css: I don't see much point for it and it'll make all HTML pages heavier for a small gain, if any.

jsha commented:
I'm happy to switch to link preload. Here's the rationale I wrote in a commit (and then forgot to hoist to the PR desc, whoops):

> This is slightly preferable to <link rel="preload" as="font" ...> because in the case that the user has the font installed locally, they can skip the network call altogether.

But it is indeed a slight benefit. I suspect most people don't have those fonts installed. I'll push a preload variant.

jsha commented:
I've updated the demo pages at https://jacob.hoffman-andrews.com/rust/flow-improvements/std/string/struct.String.html.

GuillaumeGomez commented:
My concern remains. ;)

jsha commented:
> My concern remains. ;)

As in, the `<link>` tags are unnecessary weight? Currently they are 664 bytes. The HTML for https://doc.rust-lang.org/std/string/struct.String.html is 573,186 bytes. The HTML for https://doc.rust-lang.org/std/primitive.u64.html is 525,379 bytes.

Here's why I think it's useful: Fonts are two levels deep in the page load hierarchy. The browser has to load the CSS before it even knows what URLs to fetch for the fonts. This leads to inefficient utilization of the network - the browser could have sent out requests for CSS and fonts in parallel, but now it has to load them in serial (relative to the CSS). It's a gain for the first load, though I think what you're getting at is that it may not be much of a gain for loads when the fonts are already in a disk cache?

A couple of resources. Unfortunately neither directly addresses the issue of how helpful preloading is when the resource is already on disk.

https://web.dev/preload-critical-assets/
https://web.dev/optimize-webfont-loading/

jsha commented:
Also worth mentioning #82286: static resources on doc.rust-lang.org currently don't have cache headers and get a round-trip with a 304 on every page load. That's easily fixed for doc.rust-lang.org, but other sites hosting rustdoc might have the same problem (which is a web server config problem rather than a rustdoc problem). For sites that have this problem, preloading does improve performance even if the fonts are on disk already.

Nemo157 commented:
> Add font-display: swap.

IIRC `font-display` changes were previously tried in https://github.com/rust-lang/rust/pull/72092 and reverted in https://github.com/rust-lang/rust/pull/72610, it'd be good to check whether the rational from those still apply?

GuillaumeGomez commented:
> Currently they are 664 bytes. The HTML for https://doc.rust-lang.org/std/string/struct.String.html is 573,186 bytes. The HTML for https://doc.rust-lang.org/std/primitive.u64.html is 525,379 bytes.

It's maybe "only" 664 bytes, but it has to be multiplied by the number of pages, and there are a lot.

To be noted for later: in case we decide to go on with this CSS font rule inlining, let's remove the CSS font file and simply create a function in the source code to get the CSS font rules directly. No need to keep a file around if it's only to be used at compile time.

notriddle commented:
So, in other words, something like this?

```rust
fn generates_fonts_css() -> String {
    use std::fmt::Write;
    let fonts = [
        ("Fira Sans", "normal", "400", "Fira Sans", "FiraSans-Regular.woff"),
        ("Fira Sans", "normal", "500", "Fira Sans Medium", "FiraSans-Medium.woff"),
        ("Source Serif Pro", "normal", "400", "Source Serif Pro", "SourceSerifPro-Regular.ttf.woff"),
        ("Source Serif Pro", "italic", "400", "Source Serif Pro Italic", "SourceSerifPro-It.ttf.woff"),
        ("Source Serif Pro", "normal", "700", "Source Serif Pro Bold", "SourceSerifPro-Bold.ttf.woff"),
        ("Source Code Pro", "normal", "400", "", "SourceCodePro-Regular.ttf.woff"),
        ("Source Code Pro", "normal", "600", "", "SourceCodePro-Semibold.ttf.woff"),
    ];
    let mut ret_val = "/*(C) info FONT_LICENSES.txt*/".to_string();
    for (family, style, weight, local, url) in &fonts[..] {
        write!(&mut ret_val, "@font-face{{font-family:'{family}';font-style:{style};font-weight:{weight};src:{maybe_local}url({url})}}",
            family = family,
            style = style,
            weight = weight,
            maybe_local = (if *local != "" { format!("local('{}') ", local) } else { String::new() }),
            url = url
        ).expect("writing to string is infallible");
    }
    assert!(ret_val.len() < 1000); // the exact value is 939, BTW
    ret_val
}
lazy_static!{
    static ref FONTS_CSS: String = generate_fonts_css();
}
```

We could make it even smaller by renaming the fonts (`.ttf.woff`? really?)

GuillaumeGomez commented:
@notriddle That too (I didn't look at the new `<links>` yet). I was actually referring to the new `.css` file.

jsha commented:
I pushed a new revision and demo. When I switched to `<link rel="preload">` I did an incomplete job and didn't clean up files. Sorry about that! Fixed now.

> font-display changes were previously tried in #72092 and #72557, then reverted in #72610

Thanks for the links to the history here, @Nemo157. I read through these. It looks like the first attempt was with `font-display: optional`, but that ran into issues that caused the fonts never to display. That may be related to lack of caching headers and lack of preloading, as [suggested in one of the bugzilla comments](https://bugzilla.mozilla.org/show_bug.cgi?id=1486158#c20) linked from [@workingjubilee's comment](https://github.com/rust-lang/rust/pull/72557#issuecomment-635001406).

It looks like #72557 was later updated to use `font-display: swap` instead of `font-display: optional`, but was never merged that way. I do think `font-display: swap` is more stable than `optional`, and it matches the advice on https://web.dev/font-display/#how-to-avoid-showing-invisible-text.

To be clear about possible downsides: If font loading is slow, users may see a reflow when the fonts come in. However, the status quo is that users see a blank page until the fonts come in, and then get a reflow anyhow. And preloading and cache headers should greatly reduce the likelihood of seeing a reflow.

notriddle commented:
Personally, I would rather just inline the CSS, not use preload links. It’s not that much bigger, and it’s less total code.

GuillaumeGomez commented:
@notriddle Don't have much of a preference and I'm clearly not a front-end expert. What do you think about it @Nemo157 ?

Nemo157 commented:
I personally err on the side of inlining more stuff, most of my personal websites target 1 request to load; a few extra bytes in the request that is going to happen anyway seem less expensive than doing an extra request.

bors commented:
:umbrella: The latest upstream changes (presumably #82552) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

jsha commented:
#82545 means we should really do the inlined CSS version, since `<link rel="preload">` doesn't know whether the browser supports woff or woff2. Let me know if that sounds good, and I'll update.

For the question of whether the inlined CSS should be a file, or should be a string in .rs file, or should be an `fn` that generates a string: I like the file approach since it's easier for someone to see what's in it, and edit in the plain `.css` format rather than going through a layer of indirection in `.rs`. What's the downside of making it a file?

workingjubilee commented:
> > Add font-display: swap.
> 
> IIRC `font-display` changes were previously tried in #72092 and #72557, then reverted in #72610, it'd be good to check whether the rationale from those still apply? (I don't personally know much about how the different settings interact, just bringing this up in case others that do have insights, maybe cc @workingjubilee).

As effectively already enumerated by @jsha: While adding `font-display: optional;` proved... overeager, my reasoning still held for `font-display: swap;` and `fallback` in my opinion, and I, at least, abandoned attempts to take another swing at the `font-display` CSS properties simply out of lack of energy to build consensus for them, not for any technical reason.

workingjubilee commented on src/librustdoc/html/static/rustdoc.css:
```diff
@@ -3,14 +3,18 @@
-	src: local('Fira Sans'), url("FiraSans-Regular.woff") format('woff');
+	src: local('Fira Sans'),
+		url("FiraSans-Regular.woff2") format("woff2"),
+		url("FiraSans-Regular.woff") format('woff');
```
Can you `git rebase` this so as to not include these changes as a merge commit, @jsha? rust-lang/rust uses a no-merge-commit policy (except via bors). https://rustc-dev-guide.rust-lang.org/contributing.html#opening-a-pr

I can walk you through the steps on https://rust-lang.zulipchat.com if you need help.

GuillaumeGomez commented on src/librustdoc/html/layout.rs:
```diff
@@ -55,6 +55,13 @@ crate fn render<T: Print, S: Print>(
     <link rel=\"stylesheet\" type=\"text/css\" href=\"{static_root_path}normalize{suffix}.css\">\
     <link rel=\"stylesheet\" type=\"text/css\" href=\"{static_root_path}rustdoc{suffix}.css\" \
           id=\"mainThemeStyle\">\
+    <link rel=\"preload\" as=\"font\" crossorigin=\"anonymous\" href=\"{static_root_path}FiraSans-Regular.woff2\">\
```
I really have cold feet about this change. Another reason is that currently, the css will only load the font it needs, stopping at the first one working whereas it will download all of them in any case here. So here what I suggest: can you open a PR with this change so we can go through this carefully and get the other changes merged please? Like that, we can already add most of the improvements already (which are great and I can't wait for this faster rustdoc!).

jsha replied:
Done. And I agree that the `preload` version of this conflicts with the `woff` / `woff2` change I recently landed; I think a followup change would need to do this with an inline `<style>` in order to properly account for that.

GuillaumeGomez replied:
Thanks!

jsha commented:
I pushed a small update: f9cfe15 uses Arial in place of sans-serif, which is a no-op except on Firefox, where it makes things more consistent with other platforms and reduces the size of the reflow when fonts load.

GuillaumeGomez commented:
Thanks!

@bors: r+ rollup

bors commented:
:pushpin: Commit f9cfe1583b18052bc5f42d92e8973e18ba404b19 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez f9cfe1583b18052bc5f42d92e8973e18ba404b19 -->
<!-- homu: {"type":"Approved","sha":"f9cfe1583b18052bc5f42d92e8973e18ba404b19","approver":"GuillaumeGomez"} -->

nagisa commented:
Note that removing the `sans-serif` fallback is now causing fallback to `serif` on systems that don't have `Arial` installed.

Web-safe fonts are not a thing.

jsha commented:
Thanks for the update, @nagisa! Today I also Learned the web-safe fonts are not a thing. Want to share a link or two? Also, what systems are you finding that don't have `Arial` installed?

nagisa commented:
I've a NixOS (Linux) system, though I imagine people using any "configure yourself from scratch" kind of distribution like, say, Arch Linux would be affected too.

As far as Firefox is concerned, there's a feature where users can disallow websites to select fonts other than `serif`, `sans-serif` or `monospace` entirely. In which case almost any `font-family` rule that does not specify these exact values as a fallback will end up falling back to what Firefox calls a "proportional" – which is a `serif` by default, I believe.

pickfire commented on src/librustdoc/html/static/rustdoc.css:
```diff
@@ -475,7 +482,7 @@ h4 > code, h3 > code, .invisible > code {
 #main > .since {
 	top: inherit;
-	font-family: "Fira Sans", sans-serif;
+	font-family: "Fira Sans", Arial;
```
Shouldn't we keep sans-serif as Arial may not always be available?

GuillaumeGomez replied:
That sounds logical indeed. Want to send a PR for it?

pickfire replied:
Oh, it's fixed. https://github.com/rust-lang/rust/commit/7134b0ee6721d25d2e211f01c2f344b0d8e68004

GuillaumeGomez replied:
Awesome!
